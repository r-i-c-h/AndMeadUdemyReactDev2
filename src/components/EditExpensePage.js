import React from 'react';

const EditExpensePage = (props) => {

  return (<div>EDIT Expense Page</div>);
}; 

export default EditExpensePage;