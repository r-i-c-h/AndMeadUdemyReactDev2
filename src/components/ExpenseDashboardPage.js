import React from 'react';

const ExpenseDashboardPage = () => <div>Expense Dashboard Page</div>; 

export default ExpenseDashboardPage;