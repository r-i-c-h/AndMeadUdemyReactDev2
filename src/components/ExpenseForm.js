import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

// Doesn't actually hit a reducer because the same component works for both ADD and EDIT 
// - instead /those/ pass down their own versions of props.onSubmit() 
export default class ExpenseForm extends Component {
  state = {
    amount: '',
    createdAt: moment(),
    description: '',
    note: '',
    datePickerFocused: false,
    error: ''
  };
  /** Alternate Style of onChange Handling: Instead of starting with a var defined as e.target.value **/
  /** You CAN access the e.target.value directly, BUT you MUST call e.persist() first! **/
  /* handleFooChange = (e) => { e.persist(); this.setState(()=>{ foo: e.target.value}); } */
  handleDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState( ()=> ({description}) );
  };
  handleNoteTextChange = (e) => {
    const note = e.target.value;
    this.setState( ()=> ({note}) );
  };
  handleAmountChange = (e) => {
    const amount = e.target.value;
    if ( !amount || amount.match(/^\d+(\.\d{0,2})?$/) ) {
      this.setState( ()=> ({amount}) );
    }
  };
  handleDateChange = (createdAt) => { // per docs, this gets called with a new moment.js time Obj{} as the arg...
    if (createdAt) {  // This 'if' statement ensures the field is never empty\cleared by user...
      this.setState( ()=> ({createdAt}) );
    }
  };
  onFocusChange = ({ focused }) => { // <--req'd by Air-bnb date picker component
    this.setState( ()=> ({ datePickerFocused: focused }) );
  };
  onSubmit = (e) => {
    e.preventDefault();

    if ( !this.state.description || !this.state.amount ) {
      this.setState( () => ({ error: 'Description+Amount are REQUIRED in order to Submit!' }) );
    } else {
      this.setState( () => ({ error: '' }) );
      this.props.onSubmit({ // Call the passed-in Props version so we can reuse this ExpenseForm with either ADD or EDIT
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
      { this.state.error && <p>{this.state.error}</p> }
        <form onSubmit={this.onSubmit}>
          <input type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.handleDescriptionChange}
            autoFocus
          />
          $<input type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.handleAmountChange}
          />
          <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.handleDateChange}
            focused={this.state.datePickerFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>{false}} 
          />
          <textarea
            placeholder="Note to add..."
            onChange={this.handleNoteTextChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}