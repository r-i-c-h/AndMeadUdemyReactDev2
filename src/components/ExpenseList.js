import React from 'react';
import { connect } from 'react-redux';
// connect is effectively a higher-order-component that creates a version of ExpenseList with access to the store
// connectAPI: connect( state => {Info:to access})( Component to access it with )
import selectExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = props => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No Expenses Reported In</p>
      ) : (
        props.expenses.map( exp => <ExpenseListItem key={exp.id} {...exp}/>)
      )
    }
  </div>
);

const mapStateToProps = state => {
  return { 
    expenses: selectExpenses( state.expenses, state.filters )
  };
};

export default connect(mapStateToProps)(ExpenseList);