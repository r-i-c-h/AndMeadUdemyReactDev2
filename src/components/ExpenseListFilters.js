import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, setStartDate, setEndDate, sortByAmount, sortByDate } from '../actions/filters';

class ExpenseListFilters extends Component {
  state = {
    isCalendarFocused: null
  };

  handleDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch( setStartDate(startDate) );
    this.props.dispatch( setEndDate(endDate) );
  }
  onFocusChange = (isCalendarFocused) => { 
    this.setState( () => ({isCalendarFocused}) );
  }
  
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => {
            this.props.dispatch(setTextFilter(e.target.value));
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={e => {
            if (e.target.value === 'date') {
              this.props.dispatch(sortByDate());
            } else if (e.target.value === 'amount') {
              this.props.dispatch(sortByAmount());
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.handleDatesChange}
          numberOfMonths={1}
          isOutsideRange={() => { false; }}
          showClearDates={true}
          focusedInput={this.state.isCalendarFocused}
          onFocusChange={this.onFocusChange}
        />
      </div>
    );
  }
}

const mapStateFiltersToProps = state => {
  return { filters: state.filters };
};

export default connect(mapStateFiltersToProps)(ExpenseListFilters);