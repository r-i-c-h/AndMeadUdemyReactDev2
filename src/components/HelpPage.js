import React from 'react';

const HelpPage = () => <div>HAAALLLP!</div>; 

export default HelpPage;