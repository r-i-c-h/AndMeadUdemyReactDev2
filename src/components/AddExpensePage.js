import React, { Component } from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses'; // wants Description, Note, Amount, CreatedAt

export class AddExpensePage extends Component {
  onSubmit = (expense) => {
    // /* props.dispatch(addExpense(expense)); <<< Taken care of by mapDispatchToProps */
    this.props.onSubmit(expense);
    this.props.history.push('/'); // history from `react-router` - effectively clears form by changing page
  }

  render(){
    return(
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm 
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
} 

const mapDispatchToProps = (dispatch) => {
  return { onSubmit: (expense) => dispatch(addExpense(expense)) };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);

// connect 1st arg is mapStatesToProps
// connect 2nd arg is mapDispatchToProps