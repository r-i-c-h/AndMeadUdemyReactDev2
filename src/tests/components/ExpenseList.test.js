import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import FAKE from '../FIXTURES/dummyExpensesDataArray';

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={FAKE}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList w/Msg if no Expenses', () => {
  const wrapper = shallow( <ExpenseList expenses={ [] } /> );
  expect(wrapper).toMatchSnapshot();
});

