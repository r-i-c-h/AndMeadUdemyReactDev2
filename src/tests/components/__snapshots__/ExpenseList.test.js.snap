// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList w/Msg if no Expenses 1`] = `
<div>
  <p>
    No Expenses Reported In
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={111}
    createdAt={0}
    description="thing-1"
    id="1"
    key="1"
    note="lorem ipsum 1"
  />
  <ExpenseListItem
    amount={222}
    createdAt={-345600000}
    description="the second"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={333}
    createdAt={345600000}
    description="los third-olio"
    id="3"
    key="3"
    note="threeer"
  />
</div>
`;
